### PR TITLE
Add ability to use an anvil fully

### PR DESCRIPTION
**Pull request by u9g**

### CODE REVIEW

**Comment by u9g** [comment]
doesn't work unless you set the name

**Comment by Karang** [comment]
Nice, can you add a test ?

**Comment by u9g** [comment]
Needs https://github.com/PrismarineJS/node-minecraft-protocol/issues/820 to be fixed to proceed.
this is issue is related because for an anvil, if you want to get something out of an anvil, you have to send `{name: ''}` as the entire packet, so if mineflayer can't send the packet, anvils won't work with custom names

File: lib/plugins/anvil.js [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+    async function addCustomName (name) {
+      if (!name) return
+      for (let i = 1; i < name.length; i++) {
+        sendItemName(name.substring(0, i))
+        await sleep(50)
+      }
+    }
+
+    async function combine (slotOne, slotTwo, name) {
+      if (name?.length > 35) err('Name is too long.')
```
Comment (rom1504) [inline lib/plugins/anvil.js:50]: no throw ?

File: test/externalTests/anvil.js [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+  const Item = require('prismarine-item')(bot.version)
+  const renameCost = () => mcData.isNewerOrEqualTo('1.8.9') ? 0 : 1 // weird quirk of anvils
+  const renameName = (name) => mcData.isOlderThan('1.14') ? name : JSON.stringify({ text: name }) // weird quirk of anvils
+  /* setup */
+  await bot.test.becomeCreative()
+  await bot.test.setInventorySlot(36, new Item(mcData.itemsByName.anvil.id, 1))
+  await bot.test.becomeSurvival()
+  await bot.test.placeBlock(36, bot.entity.position.offset(1, 0, 0))
+
+  if (mcData.isNewerOrEqualTo('1.13')) bot.chat(`/xp set ${bot.username} 999 levels`)
```
Comment (rom1504) [inline test/externalTests/anvil.js:14]: are you sure this is correct ?

Comment (rom1504) [reply test/externalTests/anvil.js:14]: looks like yes https://minecraft.gamepedia.com/Commands/experience

File: test/externalTests/anvil.js [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+    await bot.test.setInventorySlot(37, makeBook([{ name: 'sharpness', lvl: 5 }]))
+    await bot.test.becomeSurvival()
+
+    const anvil = await bot.openAnvil(b)
+
+    const sword = anvil.findInventoryItem(mcData.itemsByName.diamond_sword.id)
+    const book = anvil.findInventoryItem(mcData.itemsByName.enchanted_book.id)
+
+    await anvil.combine(sword, book)
+    await bot.test.wait(1000)
```
Comment (rom1504) [inline test/externalTests/anvil.js:37]: having to wait like this is a bug

File: test/externalTests/anvil.js [hunk 4]
```diff
@@ -4,7 +4,7 @@ module.exports = () => async (bot) => {
   const mcData = require('minecraft-data')(bot.version)
   const Item = require('prismarine-item')(bot.version)
   const renameCost = () => mcData.isNewerOrEqualTo('1.8.9') ? 0 : 1 // weird quirk of anvils
-  const renameName = (name) => mcData.isOlderThan('1.14') ? name : JSON.stringify({ text: name }) // weird quirk of anvils
+  const renameName = (name) => mcData.isOlderThan('1.13.2') ? name : JSON.stringify({ text: name }) // weird quirk of anvils
```
Comment (rom1504) [inline test/externalTests/anvil.js:7]: that doesn't seem to make sense
they made the change between 1.13.1 and 1.13.2 ?

File: test/externalTests/anvil.js [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+
+    await bot.test.becomeSurvival()
+
+    const anvil = await bot.openAnvil(b)
+
+    const sword = bot.inventory.slots[37]
+    const book = anvil.findInventoryItem(mcData.itemsByName.enchanted_book.id)
+
+    await anvil.combine(sword, book, 'lol')
+    await bot.test.wait(1000)
```
Comment (rom1504) [inline test/externalTests/anvil.js:101]: should be removed (replaced by an await if needed)

File: test/externalTests/anvil.js [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+
+    await bot.test.becomeSurvival()
+
+    const anvil = await bot.openAnvil(b)
+
+    const sword = bot.inventory.slots[37]
+    const book = anvil.findInventoryItem(mcData.itemsByName.enchanted_book.id)
+
+    await anvil.combine(sword, book)
+    await bot.test.wait(1000)
```
Comment (rom1504) [inline test/externalTests/anvil.js:59]: should be removed (replaced by an await if needed)

File: test/externalTests/anvil.js [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+
+    await bot.test.setInventorySlot(36, makeItem({ type: mcData.itemsByName.diamond_sword.id }))
+
+    await bot.test.becomeSurvival()
+
+    const anvil = await bot.openAnvil(b)
+
+    const sword = anvil.findInventoryItem(mcData.itemsByName.diamond_sword.id)
+    await anvil.rename(sword, 'hello')
+    await bot.test.wait(1000)
```
Comment (rom1504) [inline test/externalTests/anvil.js:78]: should be removed (replaced by an await if needed)

**Comment by rom1504** [comment]
finally working, congrats

last thing to do is removing this wait 1000

the combine function should resolve when things are done, not before, so waiting 1s should not be needed

File: test/externalTests/anvil.js (hunk 4 above)
Comment (u9g) [reply test/externalTests/anvil.js]: this works other didnt is all i can say 🤷

File: lib/plugins/inventory.js [hunk 8]
```diff
@@ -84,9 +84,8 @@ function inject (bot, { version, hideErrors }) {
       if (!window.selectedItem) {
         return
       }
-      const item = window.findItemRange(start, end, window.selectedItem.type,
-        window.selectedItem.metadata, true)
-      if (item) {
+      const item = window.findItemRange(start, end, window.selectedItem.type, window.selectedItem.metadata, true, window.selectedItem.nbt)
+      if (item && item.stackSize !== item.count) {
```
[the remark this answers is not in the document]
Comment (u9g) [reply lib/plugins/inventory.js]: I think(?) you fixed this

File: test/externalTests/anvil.js (hunk 5 above)
Comment (u9g) [reply test/externalTests/anvil.js]: fixed

File: test/externalTests/anvil.js (hunk 6 above)
Comment (u9g) [reply test/externalTests/anvil.js]: fixed

File: test/externalTests/anvil.js (hunk 7 above)
Comment (u9g) [reply test/externalTests/anvil.js]: fixed

**Review by rom1504 — APPROVED** [review]

**Pull request merged**